lighthouse-cli/test/smokehouse/smokehouse.js

```diff
@@ -1,154 +1,227 @@
+const getAssertionReport = require('./report-assert.js');
+const LocalConsole = require('./local-console.js');
+const ConcurrentMapper = require('./concurrent-mapper.js');
+
 /* eslint-disable no-console */
 
-const fs = require('fs');
-const spawnSync = require('child_process').spawnSync;
-const yargs = require('yargs');
-const log = require('lighthouse-logger');
-const rimraf = require('rimraf');
+/** @typedef {import('./lighthouse-runners/child-process-error.js')} ChildProcessError */
 
-const {collateResults, report} = require('./smokehouse-report.js');
-const assetSaver = require('../../../lighthouse-core/lib/asset-saver.js');
+// The number of concurrent (`!runSerially`) tests to run if `jobs` isn't set.
+const DEFAULT_CONCURRENT_RUNS = 5;
+const DEFAULT_RETRIES = 1;
 
-const PROTOCOL_TIMEOUT_EXIT_CODE = 67;
-const RETRIES = 3;
+/**
+ * @typedef SmokehouseOptions
+ * @property {boolean=} isDebug If true, performs extra logging from the test runs.
+ * @property {number=} jobs Manually set the number of jobs to run at once. `1` runs all tests serially.
+ * @property {number=} retries The number of times to retry failing tests before accepting. Defaults to 1.
+ * @property {LighthouseRunner=} lighthouseRunner A function that runs Lighthouse with the given options. Defaults to running Lighthouse via the CLI.
+ */
 
 /**
- * Launch Chrome and do a full Lighthouse run.
+ * @callback LighthouseRunner
  * @param {string} url
- * @param {string} configPath
- * @param {boolean=} isDebug
- * @return {Smokehouse.ExpectedRunnerResult}
+ * @param {LH.Config.Json=} configJson
+ * @param {{isDebug?: boolean}=} runnerOptions
+ * @return {Promise<{lhr: LH.Result, artifacts: LH.Artifacts, log: string}>}
  */
-function runLighthouse(url, configPath, isDebug) {
-  isDebug = isDebug || Boolean(process.env.LH_SMOKE_DEBUG);
-
-  const command = 'node';
-  const randInt = Math.round(Math.random() * 100000);
-  const outputPath = `smokehouse-${randInt}.report.json`;
-  const artifactsDirectory = `./.tmp/smokehouse-artifacts-${randInt}`;
-  const args = [
-    'lighthouse-cli/index.js',
-    url,
-    `--config-path=${configPath}`,
-    `--output-path=${outputPath}`,
-    '--output=json',
-    `-G=${artifactsDirectory}`,
-    `-A=${artifactsDirectory}`,
-    '--quiet',
-    '--port=0',
-  ];
-
-  if (process.env.APPVEYOR) {
-    // Appveyor is hella slow already, disable CPU throttling so we're not 16x slowdown
-    // see https://github.com/GoogleChrome/lighthouse/issues/4891
-    args.push('--throttling.cpuSlowdownMultiplier=1');
+
+/**
+ * Runs the selected smoke tests. Returns whether all assertions pass.
+ * @param {Array<Smokehouse.TestDfn>} smokeTestDefns
+ * @param {SmokehouseOptions=} smokehouseOptions
+ * @return {Promise<boolean>}
+ */
+async function runSmokehouse(smokeTestDefns, smokehouseOptions = {}) {
+  const {
+    isDebug,
+    jobs = DEFAULT_CONCURRENT_RUNS,
+    retries = DEFAULT_RETRIES,
```

Comment:
I don't love the default being 1 retry, for the case where you know something is failing and just want to see the diff. But I can live with it. But I think the `>0` check for this param should be `>=0`.

Reply:
It's a tension with the very parallel case, which starts failing occasionally on https://github.com/GoogleChrome/lighthouse/blob/78a1a8f8b9a8473957ce3cd491fad92936374e5c/lighthouse-cli/run.js#L160.

I'd like to move us in that direction locally to speed things up as we add more and more smoke tests (for non perf tests, it turns out even running 2x or more the number of processors isn't a big deal because most of a run doesn't come close to maxing out a core) but starting with optimizing for the local, single-run case is a reasonable starting place.

Good catch on the zero check.